parlai/core/worlds.py

```diff
@@ -550,6 +550,13 @@ def __next__(self):
 
     def batch_observe(self, index, batch_actions, index_acting):
         batch_observations = []
+        a = self.world.get_agents()[index]
+
+        if (batch_actions is not None and len(batch_actions) > 0 and
+                hasattr(a, 'batch_observe')) and index != index_acting:
+            # if there's anything that needs to be batched, do it here
+            a.batch_observe(batch_actions)
```

Comment:
the point of this function should be to give the agents the opportunity to implement a ``batch_observe`` if they prefer it over shared ``observe`` definitions, just as we do with act. in that case, we need to make sure line 571 is never called, so that needs to be contained in an else block (like we do in batch_act below--we have a "batch" case and a "call each" case).

however, as jason said, we may need to transform the observations via w.observe. Thus, we probably need to do that first to each observation, looping over the words and changing the contents of batch acts according to either calling w.observe or just validate.

Then, check for batch_observe. If it's available, call it as you do. Otherwise, call it on the particular agent as we're doing now on 571.

I think that's what we're looking for.

Here's a simple example of a batch_observe definition for the seq2seq agent, though I don't think it would help that agent since it doesn't need to pass any state around or do any batched computation in observe, so I'm not going to add it to that class for now. It might be good for us to see how you'd like to use it (since you have a use case of actual batched computation). the one advantage of this is that you wouldn't need any kind of shared definition parameters to be set up.

```
    def observe(self, observation):
        observation = copy.deepcopy(observation)
        if not self.episode_done:
            # if the last example wasn't the end of an episode, then we need to
            # recall what was said in that example
            prev_dialogue = self.observation['text']
            observation['text'] = prev_dialogue + '\n' + observation['text']
        self.observation = observation
        self.episode_done = observation['episode_done']
        return observation

    def batch_observe(self, observations):
        # observations is a list of observation dicts
        for i, obs in enumerate(observations):
            obs = copy.deepcopy(obs)
            if not self.episode_done_arr[i]:
                # if the last example wasn't the end of an episode, then we need
                # to recall what was said in that example
                prev_dialogue = self.observations[i]['text']
                obs['text'] = prev_dialogue + '\n' + obs['text']
            self.observations[i] = obs
            self.episode_done_arr[i] = obs['episode_done']
        return self.observations
```